src/functions/org/apache/jmeter/functions/TimeShift.java

```diff
@@ -0,0 +1,186 @@
+            } catch (IllegalArgumentException ex) {
+                log.error("Format date pattern '{}' is invalid (see https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html)", format, ex); // $NON-NLS-1$
+                return "";
+            }
+        }
+
+        if (!dateToShift.isEmpty()) {
+            try {
+                if (formatter != null) {
+                    localDateTimeToShift = LocalDateTime.parse(dateToShift, formatter);
+                } else {
+                    localDateTimeToShift = LocalDateTime.ofInstant(Instant.ofEpochMilli(Long.parseLong(dateToShift)),
+                            ZoneId.systemDefault());
+                }
+            } catch (DateTimeParseException | NumberFormatException ex) {
+                log.error("Failed to parse the date '{}' to shift", dateToShift, ex); // $NON-NLS-1$
+            }
+        }
+
+        // Check amount value to shift
+        if (!StringUtils.isEmpty(amountToShift)) {
+            try {
+                Duration duration = Duration.parse(amountToShift);
+                localDateTimeToShift = localDateTimeToShift.plus(duration);
+            } catch (DateTimeParseException ex) {
+                log.error("Failed to parse the amount duration '{}' to shift (see https://docs.oracle.com/javase/8/docs/api/java/time/Duration.html#parse-java.lang.CharSequence-) ", amountToShift, ex); // $NON-NLS-1$
+            }
+        }
+
+        if (formatter != null) {
+            dateString = localDateTimeToShift.format(formatter);
+        } else {
+            ZoneOffset offset = ZoneOffset.systemDefault().getRules().getOffset(localDateTimeToShift);
+            dateString = String.valueOf(localDateTimeToShift.toInstant(offset).toEpochMilli());
+        }
+
+        if (!StringUtils.isEmpty(variableName)) {
+            JMeterVariables vars = getVariables();
+            if (vars != null) {// vars will be null on TestPlan
+                vars.put(variableName, dateString);
+            }
+        }
+        return dateString;
+    }
+
+    private DateTimeFormatter createFormatter(String format) throws IllegalArgumentException {
+
+        log.debug("Create a new instance of DateTimeFormatter for format '{}' in the cache", format);
+        return new DateTimeFormatterBuilder().appendPattern(format).parseDefaulting(ChronoField.NANO_OF_SECOND, 0)
+                .parseDefaulting(ChronoField.MILLI_OF_SECOND, 0).parseDefaulting(ChronoField.SECOND_OF_MINUTE, 0)
+                .parseDefaulting(ChronoField.MINUTE_OF_HOUR, 0).parseDefaulting(ChronoField.HOUR_OF_DAY, 0)
+                .parseDefaulting(ChronoField.DAY_OF_MONTH, 1).parseDefaulting(ChronoField.MONTH_OF_YEAR, 1)
+                .parseDefaulting(ChronoField.YEAR_OF_ERA, Year.now().getValue()).toFormatter(JMeterUtils.getLocale());
+    }
+
+    protected static Cache<Object, Object> buildCache() {
```

Comment:
You can make this typesafe by returning `Cache<String, DateTimeFormatter>`. The generics are inferred at the `build()` step rather than construction, which might have caused confusion.